### PR TITLE
EVG-5938: commit queue notifications

### PR DESCRIPTION
**Pull request by ybrill**
Send notifications to users when their patch reaches the head of the queue and when the merge build passes/fails.
Implemented for both the GitHub PR and CLI commit queues.

Screen Shot 2019-04-17 at 10 43 30 AM

### CODE REVIEW

File: model/commitqueue/github_pr_sender.go [hunk 1]
```diff
@@ -6,6 +6,9 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/evergreen-ci/evergreen"
+
```
Comment (tychoish) [inline model/commitqueue/github_pr_sender.go]: nit blank line

File: model/commitqueue/github_pr_sender_test.go [hunk 2]
```diff
@@ -3,6 +3,8 @@ package commitqueue
 import (
 	"testing"
 
+	"github.com/evergreen-ci/evergreen"
+
```
Comment (tychoish) [inline model/commitqueue/github_pr_sender_test.go]: blank line

File: model/event/event_registry.go [hunk 3]
```diff
@@ -14,14 +14,12 @@ type eventRegistry struct {
...
 }
 
-var registry *eventRegistry
+var registry = &eventRegistry{
+	types:          map[string]eventDataFactory{},
+	isSubscribable: map[EventLogEntry]bool{},
+}
 
 func init() {
-	registry = &eventRegistry{
```
Comment (tychoish) [inline model/event/event_registry.go]: why this change? this might be a merge oddity?

File: model/commitqueue/github_pr_message.go [hunk 4]
```diff
@@ -18,16 +18,17 @@ const (
...
-	URL            string `bson:"url"`
-	ProjectID      string `bson:"project_id"`
-	Owner          string `bson:"owner"`
-	Repo           string `bson:"repo"`
-	Ref            string `bson:"ref"`
-	PRNum          int    `bson:"pr_number"`
-	CommitMessage  string `bson:"commit_message"`
-	CommitTitle    string `bson:"commit_title"`
-	MergeMethod    string `bson:"merge_method"`
+	PastTenseStatus string `bson:"past_tense_status"`
```
Comment (tychoish) [inline model/commitqueue/github_pr_message.go]: does changing this field to a bool break anything existing? 

The naming makes it hard to understand. Souldn't this just be Status or Outcome?

File: globals.go [hunk 5]
```diff
@@ -98,6 +98,10 @@ const (
 
 	PushStage = "push"
 
+	MergeTestStarted   = "started"
+	MergeTestSucceeded = "succeeded"
+	MergeTestFailed    = "failed"
```
[the remark this answers is not in the document]
Comment (ybrill) [reply globals.go]: Test seemed appropriate because these statuses are used when we create the patch and when the entire patch is complete, as opposed to for the individual task.

File: model/commitqueue/github_pr_sender.go (hunk 1 above)
Comment (ybrill) [reply model/commitqueue/github_pr_sender.go]: done.

File: model/commitqueue/github_pr_sender_test.go (hunk 2 above)
Comment (ybrill) [reply model/commitqueue/github_pr_sender_test.go]: done.

File: model/event/event_registry.go (hunk 3 above)
Comment (ybrill) [reply model/event/event_registry.go]: I tried to use the registry in the init function of a file that comes earlier in alphabetical order (commit_queue_event.go) and got a seg fault.

File: model/commitqueue/github_pr_message.go (hunk 4 above)
Comment (ybrill) [reply model/commitqueue/github_pr_message.go]: I think I've cleaned up all the places that were expecting a bool.
Now renamed PastTenseStatus to Status.

File: model/commitqueue/dequeue_message_test.go [hunk 6]
```diff
@@ -3,6 +3,8 @@ package commitqueue
 import (
 	"testing"
 
+	"github.com/evergreen-ci/evergreen"
+
```
Comment (tychoish) [inline model/commitqueue/dequeue_message_test.go]: nit: blank line

File: model/commitqueue/dequeue_message_test.go [hunk 7]
```diff
@@ -12,6 +14,7 @@ func TestDequeueItem(t *testing.T) {
 	dequeue := DequeueItem{
 		ProjectID: "mci",
 		Item:      "abcdef",
+		Status:    "failed",
```
Comment (tychoish) [inline model/commitqueue/dequeue_message_test.go]: use the constant?

File: model/commitqueue/github_pr_message_test.go [hunk 8]
```diff
@@ -16,6 +16,7 @@ func TestGithubPRMerge(t *testing.T) {
 		Ref:           "deadbeef",
 		CommitMessage: "merged by cq",
 		PRNum:         1,
+		Status:        "failed",
```
Comment (tychoish) [inline model/commitqueue/github_pr_message_test.go]: use constant?

File: model/commitqueue/github_pr_message_test.go [hunk 9]
```diff
@@ -49,6 +50,7 @@ func TestGithubMergePRMessageValidator(t *testing.T) {
...
```
Comment (tychoish) [inline model/commitqueue/github_pr_message_test.go]: use constant

File: model/event/commit_queue_event.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	registry.AllowSubscription(ResourceTypeCommitQueue, CommitQueueConcludeTest)
+}
+
+func commitQueueEventDataFactory() interface{} {
+	return &CommitQueueEventData{}
+}
+
+const (
+	ResourceTypeCommitQueue = "COMMIT_QUEUE"
+
```
Comment (tychoish) [inline model/event/commit_queue_event.go]: omit blank line

File: model/event/event_registry.go (hunk 3 above)
Comment (tychoish) [reply model/event/event_registry.go]: oh, I think the init() functions are being called out of order. undo this change and move the other init functions here.

Comment (ybrill) [reply model/event/event_registry.go]: done.

File: model/commitqueue/dequeue_message_test.go (hunk 6 above)
Comment (ybrill) [reply model/commitqueue/dequeue_message_test.go]: done.

File: model/commitqueue/dequeue_message_test.go (hunk 7 above)
Comment (ybrill) [reply model/commitqueue/dequeue_message_test.go]: done.

File: model/commitqueue/github_pr_message_test.go (hunk 8 above)
Comment (ybrill) [reply model/commitqueue/github_pr_message_test.go]: done.

File: model/commitqueue/github_pr_message_test.go (hunk 9 above)
Comment (ybrill) [reply model/commitqueue/github_pr_message_test.go]: done.

File: model/event/commit_queue_event.go (hunk 10 above)
Comment (ybrill) [reply model/event/commit_queue_event.go]: done.
